src/Psalm/Internal/Analyzer/Statements/Expression/CastAnalyzer.php

```diff
@@ -300,6 +289,444 @@ public static function analyze(
+                                $statements_analyzer,
+                                $return_type,
+                                $stmt,
+                                $stmt,
+                                [],
+                                $intersection_method_id,
+                                $declaring_method_id,
+                                $intersection_type->value . '::__toString',
+                                $context
+                            );
+
+                            if ($statements_analyzer->data_flow_graph) {
+                                $parent_nodes = array_merge($return_type->parent_nodes, $parent_nodes);
+                            }
+
+                            foreach ($return_type->getAtomicTypes() as $sub_atomic_type) {
+                                if ($sub_atomic_type instanceof TLiteralString) {
+                                    $valid_ints[] = new TLiteralInt((int) $sub_atomic_type->value);
+                                } elseif ($sub_atomic_type instanceof TNumericString) {
+                                    $castable_types[] = new TInt();
+                                } else {
+                                    // see above, why this is less strict
+                                    $castable_types[] = new TInt();
+                                }
+                            }
+
+                            continue 2;
+                        }
+                    }
+
+                    if ($intersection_type instanceof TObjectWithProperties
+                        && isset($intersection_type->methods['__toString'])
+                    ) {
+                        $castable_types[] = new TInt();
+
+                        continue 2;
+                    }
+                }
+            }
+
+            if ($atomic_type instanceof TNonEmptyArray
+                || $atomic_type instanceof TNonEmptyList
+            ) {
+                $risky_cast[] = $atomic_type->getId();
+
+                $valid_ints[] = new TLiteralInt(1);
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TArray
+                || $atomic_type instanceof TList
+                || $atomic_type instanceof TKeyedArray
+            ) {
+                // if type is not specific, it can be both 0 or 1, depending on whether the array has data or not
+                // welcome to off-by-one hell if that happens :-)
+                $risky_cast[] = $atomic_type->getId();
+
+                $valid_ints[] = new TLiteralInt(0);
+                $valid_ints[] = new TLiteralInt(1);
+
+                continue;
+            }
+
+            if ($atomic_type instanceof TTemplateParam) {
+                $atomic_types = array_merge($atomic_types, $atomic_type->as->getAtomicTypes());
+
+                continue;
+            }
+
+            // always 1 for "error" cases
+            $valid_ints[] = new TLiteralInt(1);
```

Comment:
Why `1` and not `int`?

Reply:
None that I can think of, but I don't have a lot of time right now. @orklah might want to do a quick double check before merging.